source/extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.cc

```diff
@@ -12,13 +12,12 @@
 
 namespace quic {
 
-QuicMemSliceImpl::QuicMemSliceImpl(QuicBufferAllocator* /*allocator*/, size_t length) {
-  Envoy::Buffer::RawSlice iovec;
-  uint64_t num_iov = single_slice_buffer_.reserve(length, &iovec, 1);
-  ASSERT(num_iov == 1);
-  // OwnedImpl may return a slice longer than needed, trim it to requested length.
-  iovec.len_ = length;
-  single_slice_buffer_.commit(&iovec, 1);
+QuicMemSliceImpl::QuicMemSliceImpl(QuicUniqueBufferPtr buffer, size_t length)
+    : buffer_(std::move(buffer)),
+      fragment_(std::make_unique<Envoy::Buffer::BufferFragmentImpl>(
+          buffer_.get(), length,
+          [](const void*, size_t, const Envoy::Buffer::BufferFragmentImpl*) {})) {
```

Comment:
It seems we don't need the fragment_ if its done callback does thing?

Reply:
This is the only way to take a given memory without copy.

Reply:
In that case, I think we can remove buffer_.

QuicMemSliceImpl::QuicMemSliceImpl(QuicUniqueBufferPtr buffer, size_t length)
    : fragment_(... buffer_.release(), ... [] (const void*p,...) { delete p; })

Reply:
Doing so requires a static_cast and const_cast of pointer p. I was trying to avoid that given that we can control the lifetime of QuicUniqueBufferPtr.

Reply:
Between a const_cast and a clearer buffer ownership, I personally think the latter is better.

Reply:
Envoy code style is to avoid unnecessary type cast, especially dynamic_cast and const_cast. And giving ownership of `buffer` to the fragment is contradict to the ideal of using BufferFragment which supposes not to own the memory. That's why the callback takes a const pointer.

Reply:
Where is the cast needed? I think it's ok to pass a non-const pointer to a function that takes a const pointer.

Having both `buffer_` and `fragment_` is not very readable to me, I can understand why a style guide wants to avoid casts, but if a cast yields more readable code I think that's the right tradeoff.

Please also consider adding a OwnedBufferFragment that encapsulates the deletion of the buffer when done. I think that's even better.

Reply:
> Where is the cast needed? I think it's ok to pass a non-const pointer to a function that takes a const pointer.
 
I thought we need to const_cast before delete, but actually not. So we don't need const_cast.

> Having both `buffer_` and `fragment_` is not very readable to me, I can understand why a style guide wants to avoid casts, but if a cast yields more readable code I think that's the right tradeoff.

Okay, I removed `buffer_` from QuicMemSliceImpl. See if it looks okay now.

> Please also consider adding a OwnedBufferFragment that encapsulates the deletion of the buffer when done. I think that's even better.